subsys/net/lib/sockets/sockets_tls.c

```diff
@@ -836,8 +889,21 @@ static int tls_mbedtls_set_credentials(struct tls_context *tls)
 exit:
 	credentials_unlock();
 
-	if (err == 0 && ca_cert_present) {
-		tls_set_ca_chain(tls);
+	if (err == 0) {
+		if (ca_cert_present) {
+			tls_set_ca_chain(tls);
+		}
+		if (own_cert_present) {
+			/* Check that the private key is set for the
+			 * leaf certificate (if any).
+			 */
+			if (!priv_key_found) {
+				err = -ENOENT;
+				goto exit;
```

Comment:
I don't think this is right, going back to exit will cause `credentials_unlock()` being called twice. Anyway, I kind of lean towards what you wrote in the related issue - I guess we could just rely on mbed TLS here to produce an error in case own certificate is registered w/o a private key (so here we'd only check for `own_cert_present`).